tests/e2e/tests/woocommerce-blocks/card-failures.spec.js

```diff
@@ -24,19 +25,34 @@ const testCard = async ( page, cardKey ) => {
 	await fillCardDetails( page, card );
 	await page.locator( 'text=Place order' ).click();
 
+	const isUpe = await isUpeCheckout( page );
+
 	/**
 	 * The invalid card error message is shown in the input field validation.
 	 * The customer isn't allowed to place the order for this type of card failure.
 	 */
-	expect
-		.soft(
-			await page.innerText(
-				cardKey === 'cards.declined-incorrect'
-					? '.wc-card-number-element .wc-block-components-validation-error'
-					: '.wc-block-checkout__payment-method .woocommerce-error'
+	if ( isUpe && cardKey === 'cards.declined-incorrect' ) {
+		expect
+			.soft(
+				await page
+					.frameLocator(
+						'.wc-block-gateway-container iframe[name^="__privateStripeFrame"]'
+					)
+					.locator( '#Field-numberError' )
+					.innerText()
+			)
+			.toMatch( new RegExp( `(?:${ card.error.join( '|' ) })`, 'i' ) );
+	} else {
+		expect
+			.soft(
+				await page.innerText(
+					cardKey === 'cards.declined-incorrect'
+						? '.wc-card-number-element .wc-block-components-validation-error'
+						: '.wc-block-checkout__payment-method .woocommerce-error'
+				)
 			)
-		)
-		.toBe( card.error );
+			.toMatch( new RegExp( `(?:${ card.error.join( '|' ) })`, 'i' ) );
+	}
```

Comment:
Nitpick: I think we could pull off the logic that's within the `soft` function into a variable before the `expect`. That would make the logic shorter and avoid code duplication. Like below:

```js
let expected;
if ( isUpe && cardKey === 'cards.declined-incorrect' ) {
	expected = await page
		.frameLocator(
			'.wc-block-gateway-container iframe[name^="__privateStripeFrame"]'
		)
		.locator( '#Field-numberError' )
		.innerText();
} else {
	expected = await page.innerText(
		cardKey === 'cards.declined-incorrect'
			? '.wc-card-number-element .wc-block-components-validation-error'
			: '.wc-block-checkout__payment-method .woocommerce-error'
	);
expect.soft( expected ).toMatch( new RegExp( `(?:${ card.error.join( '|' ) })`, 'i' ) );
```